### ISSUE TITLE
TorchInductor missing ops tracker

### ISSUE THREAD

**Issue by jansel** [pytorch/torchdynamo#327]
The following ops are using `ir.FallbackKernel`, but we should rewrite them to us decomps:
- [ ] aten._embedding_bag
- [ ] aten.nll_loss_forward
- [ ] torchvision.roi_align
- [ ] aten.grid_sampler_2d (could use the version [from here](https://github.com/pytorch/pytorch/issues/34704)
- [ ] aten.upsample_bilinear2d
- [ ] aten._cudnn_rnn (might need to hold of on this if perf doesn't match)

The following ops are needed for training support, note this is just the first failure from each benchmark so there could be more after this:
- [ ] aten._adaptive_avg_pool2d_backward
- [ ] aten.avg_pool2d_backward
- [ ] aten.expand_as
- [ ] aten._fused_moving_avg_obs_fq_helper
- [ ] aten.glu_backward
- [ ] aten.masked_fill_
- [ ] aten.max_pool2d_with_indices_backward
- [ ] aten.native_batch_norm_backward
- [ ] aten.reflection_pad2d_backward
- [ ] aten.select_scatter
- [ ] aten.convolution_backward (might need to hold of on this if perf doesn't match)

**Comment by Chillee** [comment]
> aten.convolution_backward

Unfortunately, there's no good way of decomposing this today. The gradient (wrt the weights iirc?) cannot be expressed with a forwards convolution.

> aten.nll_loss_forward

This one is WIP: https://github.com/pytorch/pytorch/pull/78491

**Comment by ezyang** [comment]
> Unfortunately, there's no good way of decomposing this today. The gradient (wrt the weights iirc?) cannot be expressed with a forwards convolution.

Well, to be more precise, if we have a suitably generalized convolution that has output shape and transposition that would be sufficient for all convolutions forwards/backwards.

**Comment by soumith** [comment]
> The gradient (wrt the weights iirc?) cannot be expressed with a forwards convolution.

It can. Where you have to pad the output image appropriately, and the convolve it with a transposed kernel weight matrix.

See the last diagram in http://soumith.ch/ex/pages/2014/08/07/why-rotate-weights-convolution-gradient/

**Comment by Chillee** [comment]
> aten.log1p

btw, this is currently a prim in Primtorch. So there won't be a further decomposition for it.

**Comment by jansel** [comment]
> > aten.expand_as
> 
> Also, this is a CompositeImplicitAutograd op - where are you seeing this today?

I am seeing it in AOT Autograd graphs for training.  I see it in multiple models for example BERT_pytorch.

**Comment by soumith** [comment]
> Well, to be more precise, if we have a suitably generalized convolution that has output shape and transposition that would be sufficient for all convolutions forwards/backwards.

A convolution that has strides, dilation and zero padding can express all needs in a closed form.
Convolution forward, backward-input, backward-weight, etc. can be expressed in terms of this convolution.

**Comment by Lezcano** [comment]
I'll have a stab at `aten._adaptive_avg_pool2d` and its backward.

**Comment by jansel** [comment]
Lowerings are more powerful, there are some things they can represent which decomps currently can't support.

Decomps are more portable, they work outside of TorchInductor and could be used to generate things like double-backwards.

If something can be done (with the same performance) both ways, a decomp is preferred, but we may need to use lowerings to support more complex stuff or in order to match the performance of eager.

"peeking twice into the data" sounds slow.  IMO we should be performance testing decomps and ensuring that they match or beat the performance of eager mode.

**Comment by Lezcano** [comment]
"peeking twice into the data" = getting an integer from a Tensor into Python-land, i.e., 2 syncs.

Also, the way I implemented it, basically implements the general `nd` version for free, but I don't know whether this is of interest. I'll post today the draft and I'll tag you guys and let you decide, that may be best.

When it comes to performance, without a fuser, eager is always going ot be fasterter than pretty much any decomp, as most kernels are implemented in one kernel, while in decompositions we are literally decomposing them into a number of kernels.

**Comment by jansel** [comment]
We can measure performance with a fuser.  Perhaps using https://github.com/pytorch/torchdynamo/pull/785

**Comment by jansel** [comment]
I updated the OP for a few items.  Looks like we still don't have a decomp/lowering for grid_sampler_2d_backward:
https://github.com/pytorch/torchdynamo/blob/cb9add9e63ea6a45716aeca89d3669fc82ab94d5/torchinductor/lowering.py#L894

You can check that file to confirm something is still missing.  Anything implemented with a fallback is in need of decomp/lowering.

**Comment by Chillee** [comment]
@Lezcano By "measuring performance", I think @jansel means "measuring performance with inductor".

For example, we benchmarked inductor performance here with the lowering: https://github.com/pytorch/torchdynamo/pull/934

**Comment by fdrocha** [comment]
I will have a go at `grid_sampler_2d_backward`

**Comment by Lezcano** [comment]
FWIW, `aten.mse_loss_backward` is  present in PyTorch (https://github.com/pytorch/pytorch/blob/b8e1c54f53404393e5fe421c26f7bfc951251e43/torch/_decomp/decompositions.py#L340). Should we mark that one as resolved?

**Comment by jansel** [comment]
> FWIW, `aten.mse_loss_backward` is present in PyTorch (https://github.com/pytorch/pytorch/blob/b8e1c54f53404393e5fe421c26f7bfc951251e43/torch/_decomp/decompositions.py#L340). Should we mark that one as resolved?

Awesome!  I think we just need to add it to this list:
https://github.com/pytorch/torchdynamo/blob/a44c28b3ccc2a3643666cf6167990730bf585aa6/torchinductor/decomposition.py#L18

**Comment by Lezcano** [comment]
I'll take `upsample_bilinear2d_backward`

**Comment by ngimel** [comment]
cc @SherlockNoMad, for things like grid_sampler_2dbackward and upsample_bilinear2d_backward we thought that a better solution would be to create infra to generate backward from the forward decomposition trace, but stop-gap direct decomposition solution is also fine I guess (we'd still need implicit grads for higher order gradients)

**Comment by SherlockNoMad** [comment]
I added a decomp for aten.im2col. 
Also I noticed that nn.functional.unfold would eventually call aten.im2col, so I am wondering where are we seeing aten.unfold? 

```
# Module stack: {'self_network_0': 'Sequential', 'self_network_0_2': 'Outlooker', 'self_network_0__2__attn': 'OutlookAttention', 'self_network_0__2__attn_unfold': 'Unfold'}, File: /fsx/users/bahuang/conda/envs/pt_dev/lib/python3.9/site-packages/timm/models/volo.py:111, code: v = self.unfold(v).reshape(
im2col_default_2 = torch.ops.aten.im2col.default(permute_default_15, [3, 3], [1, 1], [1, 1], [2, 2]);  permute_default_15 = None
```

**Comment by SherlockNoMad** [comment]
aten._fused_moving_avg_obs_fq_helper is showing up in mobilenet_v2_quantized_qat and resnet50_quantized_qat models. 

These are quantization model, and the op is originated from FusedMovingAvgObsFakeQuantize module. Does inductor, or PT2, want to cover quantization model? I thought no... 

```
# Module stack: {'self_activation_post_process_0': 'FusedMovingAvgObsFakeQuantize'}, File: <eval_with_key>.2:5, code: activation_post_process_0 = self.activation_post_process_0(x);  x = None
_fused_moving_avg_obs_fq_helper_default = torch.ops.aten._fused_moving_avg_obs_fq_helper.default(primals_1075, primals_159, primals_160, primals_163, primals_164, primals_161, primals_162, 0.01, 0, 127, -1);  primals_1075 = primals_159 = primals_160 = primals_163 = primals_164 = primals_161 = primals_162 = None
getitem = _fused_moving_avg_obs_fq_helper_default[0];  _fused_moving_avg_obs_fq_helper_default = None
```

**Comment by Chillee** [comment]
@SherlockNoMad I think `torch.Tensor.unfold` calls `aten::unfold` (https://pytorch.org/docs/stable/generated/torch.Tensor.unfold.html), while the functional version calls `im2col`: https://pytorch.org/docs/stable/generated/torch.nn.Unfold.html

Any idea why this is the case? maybe @ngimel?

**Comment by ngimel** [comment]
Sorry, I don't know, are they semantically different?

**Comment by Lezcano** [comment]
Might be interesting to remove the `_backward` functions from high prio, as they'll hopefully be implemented automatically from the prims?

**Comment by jansel** [comment]
> Might be interesting to remove the `_backward` functions from high prio, as they'll hopefully be implemented automatically from the prims?

AFAIK generating backwards automatically is not working yet.

**Comment by Chillee** [comment]
@jansel We also were not able to obtain good performance from the automatically generated backwards for upsample_bilinear2d.

**Comment by Lezcano** [comment]
As discussed in the catch-up, I would like to understand better the reason as to why we weren't able to get good performance out of the generated `upsample_bilinear2d_backward`.

In any case, my point is that, if we are going for coverage, if we (@SherlockNoMad was doing it?) implement the ideas on how to generate backward formulas from the forward graphs, we could knock out almost all the remaining things. Then, we could concentrate in figuring out whether we want/need to implement some of these manually, or the automatic implementation is already good enough.

**Comment by jansel** [comment]
We need performant versions of backwards ops.  If the auto-generated ones aren't fast enough, the  we will need to write them by hand.

**Comment by Lezcano** [comment]
But we don't know whether the auto-generated versions aren't fast enough a priori. I'd be surprised if the autogenerated version of `grid_sampler_2d_backward` weren't performant, for example.

**Comment by Chillee** [comment]
@Lezcano That's easy to test today, similar to how I tested that the autogenerated `upsample_bilinear2d` backwards wasn't performant.

Just do something like...
```
def f(x):
    return decomposition_table[aten.grid_sampler_2d](x).sum().backward()

fx_g = make_fx(f)(torch.randn(..., requires_grad=True)
```

And then benchmark it compared to eager-mode, at least.

**Comment by Lezcano** [comment]
Would it be fair to cross out from this list those operations that we find that are at least as fast as eager?
Also, that `f` does a `sum` which is a reduction, and thus cannot be fused. Would it be fair to do
```python
torch.autograd.grad([decomposition_table[aten.grid_sampler_2d](x)], [x], grad_outputs=[g])
```
where `g = torch.ones_like(x)` is precomputed?

**Comment by fdrocha** [comment]
I will work on `upsample_bicubic2d`

**Comment by Lezcano** [comment]
I'm doing `index_{add,copy}`.

**Comment by desertfire** [comment]
> I'm doing `index_{add,copy}`.

Thanks! I have assigned https://github.com/pytorch/torchdynamo/issues/885 to you for tracking purpose.

**Comment by Chillee** [comment]
@Lezcano I don't see any reason why a sum can't be fused? Also, tbh, not fusing it would be a fairer comparison to eager, as usually your fw + bw graph need to be broken up.

**Comment by ngimel** [comment]
@lezcano, `sum` can be and is fused, but that's a good point, in backward instead of full-size gradOutput we are getting expanded single-element tensor, which is less indicative of the real performance, so the benchmarking script has to be minimally changed
```
def f(x, gO):
    out = decomposition_table[aten.grid_sampler_2d](x)
    out.backward(gO)
    return x.grad
    

fx_g = make_fx(f)(torch.randn(..., requires_grad=True), torch.randn(...))
```

**Comment by Lezcano** [comment]
I thought someone mentioned that reductions cannot be fused with each other. In any case, yes, I think the cleanest way to do this is the one proposed by Natalia.

**Comment by lezcano** [comment]
I'll take `col2im` (and probably after that `unfold`)

**Comment by fdrocha** [comment]
I'll also take `softplus` and its backward

**Comment by min-jean-cho** [comment]
Hi @jansel, I will add `aten._uniform` decompose -- please refer to #90815 . Please help let me know if any comments on this, thanks!

**Comment by jansel** [comment]
> Hi @jansel, I will add `aten._uniform` decompose -- please refer to [pytorch/pytorch#90815](https://github.com/pytorch/pytorch/issues/90815). Please help let me know if any comments on this, thanks!

Sounds good to me, thanks!

**Comment by min-jean-cho** [comment]
Hi @jansel, I think we can also decompose `aten._normal` with `torch.rand`, `torch.log`, etc similar to how numpy implements it https://github.com/numpy/numpy/blob/main/numpy/random/src/legacy/legacy-distributions.c#L18-L40 based on this method https://en.wikipedia.org/wiki/Marsaglia_polar_method.

Once `aten._normal` is decomposed, other distributions can be pretty simply decomposed as well based on the decomposed `aten._normal`. 

I wanted to ask if you have any comments on this before digging deeper, thanks!

**Comment by lezcano** [comment]
Note that these iterative methods will be particularly inefficient on GPUs. I get the feeling that many of these sampling ops should be primitives in PrimTorch. One rule of thumb we chose when writing what's a primitive for PrimTorch and what's not is whether the STL has an exclusive operation for it. If it does, it makes sense that that operation may have something numerically particular that it does not make it amenable to being decomposed.

Note that I bring up PrimTorch here even though Inductor does not necessarily abide by it, as I don't think these decompositions are specific to Inductor, but could be used by nvfuser or other consumers really.

cc @mruberry

**Comment by jgong5** [comment]
> If it does, it makes sense that that operation may have something numerically particular that it does not make it amenable to being decomposed.

Numeric is one aspect. The other is about the tradeoff between narrowed opset to support vs. performance - narrowed opset reduces the complexity of the backend support but it might be harder for backend to map the decomposed pattern into a more efficient implementation. Good to know that STL was used as one of the rules for defining the opset. Is there any documentation on the guidelines of defining the Prim opset?

**Comment by ngimel** [comment]
`randn` already has a lowering that can be reused for `_normal`. `randn` is already a primitive op in inductor, with codegen defined for cpu and triton.

**Comment by min-jean-cho** [comment]
> Hi @jansel, I think we can also decompose aten._normal


Hi all, I know there's an ongoing discussion of Prim Op --vs-- decomposed Aten Op regarding `aten.uniform_` , but just wanted to add here that I've created a RFC for decomposing `aten.normal_` as of now, https://github.com/pytorch/pytorch/issues/91085

**Comment by ngimel** [comment]
Given that backends typically provide `randn` implementation, I don't think `normal` decomposition using polar method is needed, normal should just call `randn`
